certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,742 @@
+    require e.msg.value == 0;                       // view functions revert if you send eth
+
+    getCurrentFrame@withrevert(e);
+    assert !lastReverted;
+}
+
+
+//  2. Only ACL roles can call the external non-view functions:
+//      setFrameConfig() - MANAGE_FRAME_CONFIG_ROLE 
+//      setFastLaneLengthSlots() - MANAGE_FAST_LANE_CONFIG_ROLE
+//      addMember() - MANAGE_MEMBERS_AND_QUORUM_ROLE
+//      removeMember() - MANAGE_MEMBERS_AND_QUORUM_ROLE
+//      setQuorum() - MANAGE_MEMBERS_AND_QUORUM_ROLE, DISABLE_CONSENSUS_ROLE (only if disabling it)
+//      disableConsensus() - MANAGE_MEMBERS_AND_QUORUM_ROLE, DISABLE_CONSENSUS_ROLE
+//      setReportProcessor() - MANAGE_REPORT_PROCESSOR_ROLE
+//      submitReport() - only if getIsMember(msg.sender) == true
+//      updateInitialEpoch() - DEFAULT_ADMIN_ROLE
+// Status: Pass
+// https://prover.certora.com/output/80942/42f639882bbe4b94bf5b0126311013b2/?anonymousKey=8702698b3953d812f76b895bcc3878460864d7a6
+rule onlyAllowedRoleCanCallMethod(method f) 
+    filtered { f -> !f.isView }
+{
+    env e; calldataarg args; env e2;
+
+    bool hasRole_MANAGE_FRAME_CONFIG_ROLE       = hasRole(e,MANAGE_FRAME_CONFIG_ROLE(),e.msg.sender);
+    bool hasRole_MANAGE_FAST_LANE_CONFIG_ROLE   = hasRole(e,MANAGE_FAST_LANE_CONFIG_ROLE(),e.msg.sender);
+    bool hasRole_MANAGE_MEMBERS_AND_QUORUM_ROLE = hasRole(e,MANAGE_MEMBERS_AND_QUORUM_ROLE(),e.msg.sender);
+    bool hasRole_DISABLE_CONSENSUS_ROLE         = hasRole(e,DISABLE_CONSENSUS_ROLE(),e.msg.sender);
+    bool hasRole_MANAGE_REPORT_PROCESSOR_ROLE   = hasRole(e,MANAGE_REPORT_PROCESSOR_ROLE(),e.msg.sender);
+    bool hasRole_DEFAULT_ADMIN_ROLE             = hasRole(e,DEFAULT_ADMIN_ROLE(),e.msg.sender);
+
+    bool isMember = getIsMember(e,e.msg.sender);
+
+    uint256 quorumBefore = getQuorum(e);
+
+    f@withrevert(e,args);
+    bool callReverted = lastReverted;
+
+    uint256 quorumAfter = getQuorum(e2);
+
+    assert ((f.selector == setFrameConfig(uint256,uint256).selector) &&
+            !hasRole_MANAGE_FRAME_CONFIG_ROLE) => callReverted;
+    
+    assert ((f.selector == setFastLaneLengthSlots(uint256).selector) &&
+            !hasRole_MANAGE_FAST_LANE_CONFIG_ROLE) => callReverted;
+    
+    assert ((f.selector == addMember(address,uint256).selector) &&
+            !hasRole_MANAGE_MEMBERS_AND_QUORUM_ROLE) => callReverted;
+    
+    assert ((f.selector == removeMember(address,uint256).selector) &&
+            !hasRole_MANAGE_MEMBERS_AND_QUORUM_ROLE) => callReverted;
+
+    assert ((f.selector == setQuorum(uint256).selector) &&
+            !hasRole_MANAGE_MEMBERS_AND_QUORUM_ROLE &&                          // without permissions, you
+            !hasRole_DISABLE_CONSENSUS_ROLE) => (quorumBefore == quorumAfter);  // cannot change the quorum
+    
+    assert ((f.selector == disableConsensus().selector) && 
+            quorumBefore != UNREACHABLE_QUORUM() &&         // this ensures the quorum was not disabled already
```

Comment:
I would expect that the call will revert regardless of the quorum state before, as only the role member can perform the transaction.

Reply:
I also thought this way, but they have a weird implementation here. They don't check the required role until a change is about to be made.

So, effectively, anyone can call the function if it does not change the state.